### PR TITLE
Political Algorithm ready to draft and test

### PR DESCRIPTION
**Pull request by ugomuhieddine17**
Closes # 

## 📑 Description

The purpose of the pr is to introduce the PO algorithm to the draft repo. The PO.py and the run_PO.py files have been created.

## ✅ Checks

- [] My pull request adheres to the code style of this project
- [] My code requires changes to the documentation
- [] I have updated the documentation as required
- [] All the tests have passed

### CODE REVIEW

File: models/multiple_solution/human_based/PO.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+                # Implementation of eq 10
+                if ((pjitminus1[k] <= pji[k] <= mstar) or (pjitminus1[k] >= pji[k] >= mstar)):
+                    pji[k] = mstar + (2 * r - 1) * abs(mstar - pji[k])
+                elif ((pjitminus1[k] <= mstar <= pji[k]) or (pjitminus1[k] >= mstar >= pji[k])):
+                    pji[k] = pjitminus1[k] + r * (pji[k] - pjitminus1[k])
+                elif ((mstar <= pjitminus1[k] <= pji[k]) or (mstar >= pjitminus1[k] >= pji[k])):
+                    pji[k] = mstar + (2 * r - 1) * abs(mstar - pjitminus1[k])
+                mstar = cstarj[k]
+                r = np.random.rand()
+                # Copy paste of what was above, since eq 10 already implemented
```
[the remark this answers is not in the document]
Comment (thieu1995) [reply models/multiple_solution/human_based/PO.py:59]: @ugomuhieddine17 can you see all of my comments?

**Review by thieu1995 — CHANGES_REQUESTED** [review]

File: models/multiple_solution/human_based/PO.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+        #print(f'The selected party leaders: {party_leaders}')
+
+        ################
+        # Computing the constituency leaders, 1 member for each party competes against each other
+        # So in our toy example it is individual 0 vs. 3 vs. 6
+        # In general it is 0 * n, 1 * n ...
+        ################
+
+        # first we make the constituencies
```
[the remark this answers is not in the document]
Comment (ugomuhieddine17) [reply models/multiple_solution/human_based/PO.py:158]: Each member is part of a party, and part of a constituency. It is just two different social organizations (one is a political group, and the second is the city council for example)

File: models/multiple_solution/human_based/PO.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        self.fun = po_params['fun']
+        self.lambdamax = po_params['lambdamax']
+        self.n = po_params['n']
+        self.tmax = po_params['tmax']
+        self.lambdamax = po_params['lambdamax']
+        self.d = po_params['d']
+
+        
+    def ElectionCampaign(self, pji, pjitminus1, pstari, cstarj):
+        if (self.fun(pji) <= self.fun(pjitminus1)):
```
[the remark this answers is not in the document]
Comment (ugomuhieddine17) [reply models/multiple_solution/human_based/PO.py]: done !

File: models/multiple_solution/human_based/PO.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ugomuhieddine17) [reply models/multiple_solution/human_based/PO.py:59]: it is a relict when we were coding altogether

File: models/multiple_solution/human_based/PO.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+                if ((pjitminus1[k] <= pji[k] <= mstar) or (pjitminus1[k] >= pji[k] >= mstar)):
+                    pji[k] = mstar + (2 * r - 1) * abs(mstar - pji[k])
+                elif ((pjitminus1[k] <= mstar <= pji[k]) or (pjitminus1[k] >= mstar >= pji[k])):
+                    pji[k] = pjitminus1[k] + r * (pji[k] - pjitminus1[k])
+                elif ((mstar <= pjitminus1[k] <= pji[k]) or (mstar >= pjitminus1[k] >= pji[k])):
+                    pji[k] = mstar + (2 * r - 1) * abs(mstar - pjitminus1[k])
+        return pji
+
+    def PartySwitching(self, parties, population, Lambda, fitnesst):
+        for i in range(len(parties)):
```
[the remark this answers is not in the document]
Comment (ugomuhieddine17) [reply models/multiple_solution/human_based/PO.py:69]: should be good now !

File: models/multiple_solution/human_based/PO.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+            # got the ith party
+            for j in range(len(parties[i])):
+                # for each member of the party
+                sp = np.random.rand()
+                if (sp < Lambda):
+                    r = np.random.randint(0, self.n - 1)  # the random new party (from 0 to n-1 as it how python indexes)
+                    q = np.argmax(fitnesst[parties[r]])  # do we have to recompute here? or time consuming
```
[the remark this answers is not in the document]
Comment (ugomuhieddine17) [reply models/multiple_solution/human_based/PO.py]: relicat again

File: models/multiple_solution/human_based/PO.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+            if fitness_new <= fitness_j:
+                population[int(constituencyLeaders[j])] = c_new
+
+        return 0
+
+    def _train__(self):
+       
+        ################
+        # Initialization: sth arbitrary now to replicate the paper, making 9 2D individuals
+        # n = 3 parties, d = 2 dimensions
```
[the remark this answers is not in the document]
Comment (ugomuhieddine17) [reply models/multiple_solution/human_based/PO.py:116]: it is a relicat too, we indeed generalised n and d

File: models/multiple_solution/human_based/PO.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+        ################
+        # Starting while loop
+        ################
+        self.df = pd.DataFrame()
+        while (t <= self.tmax):
+            populationtemp = populationt.copy()
+
+            fitnesstemp = fitnesst.copy()
+            for i in range(len(parties)):
+                for j in range(len(party)):
```
[the remark this answers is not in the document]
Comment (ugomuhieddine17) [reply models/multiple_solution/human_based/PO.py:195]: done !

[the remark this answers is not in the document]
Comment (ugomuhieddine17) [reply models/multiple_solution/human_based/PO.py:195]: should I cancel this pull request and do a new one to be able to see the modifications?

[the remark this answers is not in the document]
Comment (thieu1995) [reply models/multiple_solution/human_based/PO.py:195]: @ugomuhieddine17 No need to do it.

**Pull request merged**